src/server.c

```diff
@@ -5939,7 +5956,8 @@ sds genRedisInfoString(dict *section_dict, int all_sections, int everything) {
             "eventloop_duration_sum:%llu\r\n"
             "eventloop_duration_cmd_sum:%llu\r\n"
             "instantaneous_eventloop_cycles_per_sec:%llu\r\n"
-            "instantaneous_eventloop_duration_usec:%llu\r\n",
+            "instantaneous_eventloop_duration_usec:%llu\r\n"
+            "pipeline_average_last_sec:%.2f\r\n",
```

Comment:
We added a similar metric internally in AWS as well, but ultimately decided we didn't think it made sense for the community. The reason was that it's visible indirectly from other metrics such as number of read calls or time spent doing reads, since you can compare that against the total number of commands, so it wasn't clear exactly what this metric was doing. We were also concerned this metric wasn't actually "pipeline" depth, because there might be other reasons such as TCP fragmentation that prevents that actual pipeline depth from being reached.

Another concern I have is that I don't really like maintaining that many "instantaneous metrics", since they're really only useful for debugging. I would rather be emitting the "read call" count and then having some third party aggregator doing the math to figure out the "average number of commands per read call".

The last major drawback is it's adding code on the hotpath, and unless you have a very strong reason to do that would rather we not.

Reply:
Thank you for your detailed response,  these points you have raised have also been discussed within my colleagues. I will share my views in more detail.

> it's visible indirectly from other metrics such as number of read calls or time spent doing reads

I agree. These indirect indicators can obtain approximate results, but they are not precise enough. The main reason is that it is impossible to exclude requests from the master client, as well as requests from lua/module/function.

> there might be other reasons such as TCP fragmentation that prevents that actual pipeline depth from being reached

This is also our main concern. I think pipeline itself is the bulk request state observed on the server side, so I still think it makes sense, for example, in one scenario: The client is always pipeline 100, but because the network is fragmented in transit, the server side observes that the pipeline value is 50, which can be used to observe the change of the network state, or guide the user to reduce the amount of pipeline to obtain the best delay and throughput.

In one of our scenarios, the user sends requests to our proxy, which will try its best to assemble a pipeline and send it to redis. We need to observe and measure the relationship between pipeline, cpu and latency. The Lettuce6 client also uses pipeline instead of connection pool, the user does not know how many requests are sent in a batch, and also requires the server to obtain detailed information.

Reply:
i think you can get much insight from the new metrics added in #11963
e.g. you can know the average number of commands processed in each event loop.
i.e. delta of `total_commands_processed` divided by delta of `eventloop_cycles`.

it doesn't tell you anything per client, but it might even be better.
i.e. your argument about adding more pipeline creating just latency without improving the throughput, is also true for adding more clients.

Reply:
@oranagra thanks, I had read the new metrics about eventloop. Thanks to contributors, I think it's useful in throubleshooting. But I don't know how to use these metrics to guide improving the throughput of redis. Through the benchmark tool, I find that the performance of pipeline=2 is almost twice as good as that of pipeline=1, with almost no change in latency. 
For example, if I can clearly know that the current redis side pipeline is 1 through this indicator, then I can make clear performance optimization suggestions to the user and recommendations the user to upgrade the pipeline to 2